File: app/src/main/java/com/boardgamegeek/filterer/CollectionNameFilter.java
package com.boardgamegeek.filterer;

import android.content.Context;
import android.support.annotation.NonNull;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Collection;

public class CollectionNameFilter extends CollectionFilterer {
	private String filterText;
	private boolean startsWith;

	public CollectionNameFilter(Context context) {
		super(context);
	}

	public CollectionNameFilter(Context context, CharSequence filterText, boolean startsWith) {
		super(context);
		this.filterText = filterText.toString();
		this.startsWith = startsWith;
	}

	@Override
	public void setData(@NonNull String data) {
		int lastIndex = data.lastIndexOf(DELIMITER);
		if (lastIndex == -1) {
			filterText = data;
			startsWith = false;
		} else {
			filterText = data.substring(0, lastIndex);
			startsWith = data.substring(lastIndex).equals("1");
		}
	}

	@Override
	public int getTypeResourceId() {
		return R.string.collection_filter_type_collection_name;
	}

	@Override
	public String getDisplayText() {
		if (startsWith) return filterText + "*";
		return "*" + filterText + "*";
	}

	@Override
	public String getSelection() {
		return Collection.COLLECTION_NAME + " LIKE ?";
	}

	@Override
	public String[] getSelectionArgs() {
		if (startsWith) return new String[] { filterText + "%" };
		return new String[] { "%" + filterText + "%" };
	}

	@Override
	public String flatten() {
		return filterText + DELIMITER + (startsWith ? "1" : "0");
	}

	public String getFilterText() {
		return filterText;
	}

	public boolean startsWith() {
		return startsWith;
	}
}

File: app/src/main/java/com/boardgamegeek/filterer/CollectionNameFilter.kt
package com.boardgamegeek.filterer

import android.content.Context
import com.boardgamegeek.R
import com.boardgamegeek.model.CollectionItem

class CollectionNameFilter(context: Context) : CollectionFilterer(context) {
    var filterText = ""
    var startsWith = false

    override val typeResourceId = R.string.collection_filter_type_collection_name

    override fun inflate(data: String) {
        filterText = data.substringBeforeLast(DELIMITER)
        startsWith = data.substringAfterLast(DELIMITER) == "1"
    }

    override fun deflate() = "$filterText$DELIMITER${if (startsWith) "1" else "0"}"

    override val iconResourceId: Int
        get() = R.drawable.ic_baseline_format_quote_24

    override fun chipText() = if (startsWith) "$filterText*" else "*$filterText*"

    override fun description() = context.getString(if (startsWith) R.string.starts_with_prefix else R.string.named_prefix, filterText)

    override fun filter(item: CollectionItem): Boolean {
        return if (startsWith) {
            item.collectionName.startsWith(filterText, true)
        } else {
            item.collectionName.contains(filterText, true)
        }
    }
}
